shark-cli/src/main/java/shark/Neo4JCommand.kt

```diff
@@ -0,0 +1,700 @@
+              )
+            )
+          }
+        }
+        echo("Progress labels: 100%, committing transaction")
+        labelsTx.commit()
+
+        val gcRootsTx = dbService.beginTx()
+        echo("Progress gc roots: 0%")
+        lastPct = 0
+        val gcRootTotal = graph.gcRoots.size
+
+        // A root for all gc roots that makes it easy to query starting from that single root.
+        gcRootsTx.execute("create (:GcRoots {name:\"GC roots\", leakingStatus:${NOT_LEAKING.name}})")
+
+        graph.gcRoots.forEachIndexed { index, gcRoot ->
+          val pct = ((index * 10f) / gcRootTotal).toInt()
+          if (pct != lastPct) {
+            lastPct = pct
+            echo("Progress gc roots: ${pct * 10}%")
+          }
+          gcRootsTx.execute(
+            "match (roots:GcRoots), (object:Object{objectId:\$objectId}) create (roots)-[:ROOT]->(:GcRoot {type:\$type})-[:ROOT]->(object)",
+            mapOf(
+              "objectId" to gcRoot.id,
+              "type" to gcRoot::class.java.simpleName
+            )
+          )
+        }
+        gcRootsTx.commit()
+      }
+
+      echo("Retrieving server bolt port...")
+
+      // TODO Unclear why we need to query the port again?
+      val boltPort = dbService.executeTransactionally(
+        "CALL dbms.listConfig() yield name, value " +
+          "WHERE name = 'dbms.connector.bolt.listen_address' " +
+          "RETURN value", mapOf()
+      ) { result ->
+        val listenAddress = result.next()["value"] as String
+        val pattern = Pattern.compile("(?:\\w+:)?(\\d+)")
+        val matcher = pattern.matcher(listenAddress)
+        if (!matcher.matches()) {
+          error("Could not extract bolt port from [$listenAddress]");
+        }
+        matcher.toMatchResult().group(1)
+      }
+
+      val browserUrl = "http://browser.graphapp.io/?dbms=bolt://localhost:$boltPort"
+      echo("Opening: $browserUrl")
+      Runtime.getRuntime().exec("open $browserUrl")
+      ConsoleReader().readLine("Press ENTER to shut down Neo4j server")
+      echo("Shutting down...")
+      managementService.shutdown()
+    }
+
+    fun HeapValue.heapValueAsString(): String {
+      return when (val heapValue = holder) {
+        is ReferenceHolder -> {
+          if (isNullReference) {
+            "null"
+          } else {
+            error("should not happen")
+          }
+        }
+        is BooleanHolder -> heapValue.value.toString()
+        is CharHolder -> heapValue.value.toString()
+        is FloatHolder -> heapValue.value.toString()
+        is DoubleHolder -> heapValue.value.toString()
+        is ByteHolder -> heapValue.value.toString()
+        is ShortHolder -> heapValue.value.toString()
+        is IntHolder -> heapValue.value.toString()
+        is LongHolder -> heapValue.value.toString()
+      }
+    }
+  }
+}
+
+class FindLeakPaths {
+
+  @org.neo4j.procedure.Context
+  lateinit var transaction: Transaction
+
+  @UserFunction("shark.leakPaths")
```

Comment:
note for later: this is semantically closer to a procedure, than a user function